subsys/mpsl/cx/thread/mpsl_cx_thread.c

```diff
@@ -76,33 +79,57 @@ static int32_t granted_ops_get(mpsl_cx_op_map_t *granted_ops)
 	return 0;
 }
 
+static void check_and_notify(void)
+{
+	int32_t ret;
+	mpsl_cx_op_map_t granted_ops;
+
+	ret = granted_ops_get(&granted_ops);
+
+	__ASSERT(ret == 0, "Getting grant pin state returned unexpected result: %d", ret);
+	if (ret != 0) {
+		/* nrfx gpio implementation cant return failure for this call
+		 * This condition is handled for fail-safe approach. It is
+		 * assumed GRANT is not given, if cannot read its value
+		 */
+		granted_ops = granted_ops_map(false);
+	}
+
+	if (granted_ops != last_notified) {
+		last_notified = granted_ops;
+		callback(granted_ops);
+	}
+}
+
 static void gpiote_irq_handler(const struct device *gpiob, struct gpio_callback *cb, uint32_t pins)
 {
 	(void)gpiob;
 	(void)cb;
 	(void)pins;
+	uint8_t mon;
 
-	static mpsl_cx_op_map_t last_notified;
-	int32_t ret;
-	mpsl_cx_op_map_t granted_ops;
-
-	if (callback != NULL) {
-		ret = granted_ops_get(&granted_ops);
-
-		__ASSERT(ret == 0, "Getting grant pin state returned unexpected result: %d", ret);
-		if (ret != 0) {
-			/* nrfx gpio implementation cannot return failure for this call
-			 * This condition is handled for fail-safe approach. It is assumed
-			 * GRANT is not given, if cannot read its value
-			 */
-			granted_ops = granted_ops_map(false);
-		}
+	if (callback == NULL) {
+		return;
+	}
 
-		if (granted_ops != last_notified) {
-			last_notified = granted_ops;
-			callback(granted_ops);
+	do {
+		if (is_handled == 0) {
+			is_handled++;
+			check_and_notify();
+			is_handled--;
```

Comment:
I see a potential race here. Let's imagine this function is executed from a thread context an it is preempted by an interrupt just before `is_handled` is decremented. The interrupt handler will skip to `else` and increment the monitor. The thread will execute exclusive load-store loop decrementing the monitor value. The outer loop won't iterate again, but it should.
Moving the decrementing loop before `check_and_notify()` should fix this issue.